src/main/scala/strawman/collection/mutable/ListBuffer.scala

```diff
@@ -33,29 +41,163 @@ class ListBuffer[A]
+    if (p == null) first else p.next
+
+  private def setNext(p: Predecessor[A], nx: List[A]): Unit =
+    if (p == null) first = nx else p.next = nx
+
+  def update(idx: Int, elem: A): Unit = {
+    ensureUnaliased()
+    if (idx < 0 || idx >= len) throw new IndexOutOfBoundsException
+    val p = locate(idx)
+    setNext(p, elem :: getNext(p).tail)
+  }
+
+  def insert(idx: Int, elem: A): Unit = {
+    ensureUnaliased()
+    if (idx < 0 || idx > len) throw new IndexOutOfBoundsException
+    if (idx == len) +=(elem)
+    else {
+      val p = locate(idx)
+      setNext(p, elem :: getNext(p))
+      len += 1
+    }
+  }
+
+  private def insertAfter(p: Predecessor[A], it: Iterator[A]) = {
+    var prev = p
+    val follow = getNext(prev)
+    while (it.hasNext) {
+      len += 1
+      val next = (it.next :: follow).asInstanceOf[::[A]]
+      setNext(prev, next)
+      prev = next
+    }
+  }
+
+  def insertAll(idx: Int, elems: IterableOnce[A]): Unit = {
+    ensureUnaliased()
+    val it = elems.iterator()
+    if (it.hasNext) {
+      ensureUnaliased()
+      if (idx < 0 || idx > len) throw new IndexOutOfBoundsException
+      if (idx == len) ++=(elems)
+      else insertAfter(locate(idx), it)
+    }
+  }
+
+  def remove(idx: Int): A = {
+    ensureUnaliased()
+    if (idx < 0 || idx >= len) throw new IndexOutOfBoundsException
+    len -= 1
+    val p = locate(idx)
+    val nx = getNext(p)
+    setNext(p, nx.tail)
+    nx.head
+  }
+
+  def remove(idx: Int, n: Int): Unit =
+    if (n > 0) {
+      ensureUnaliased()
+      if (idx < 0 || idx + n > len) throw new IndexOutOfBoundsException
+      removeAfter(locate(idx), n)
+    }
+
+  private def removeAfter(prev: Predecessor[A], n: Int) = {
+    @tailrec def ahead(p: List[A], n: Int): List[A] =
+      if (n == 0) p else ahead(p.tail, n - 1)
+    setNext(prev, ahead(getNext(prev), n))
+    len -= n
+  }
+
+  def mapInPlace(f: A => A): this.type = {
+    ensureUnaliased()
+    val buf = new ListBuffer[A]
+    for (elem <- this) buf += f(elem)
+    first = buf.first
+    last = buf.last
+    this
+  }
+
+  def flatMapInPlace(f: A => IterableOnce[A]): this.type = {
+    ensureUnaliased()
+    val prev: Predecessor[A] = null
+    var cur: List[A] = first
+    while (!cur.isEmpty) {
+      val follow = cur.tail
+      setNext(prev, follow)
+      len -= 1
+      insertAfter(prev, f(cur.head).iterator())
+      cur = follow
+    }
+    this
+  }
+
+  def filterInPlace(p: A => Boolean): this.type = {
+    ensureUnaliased()
+    var prev: Predecessor[A] = null
+    var cur: List[A] = first
+    while (!cur.isEmpty) {
+      val follow = cur.tail
+      if (!p(cur.head)) {
+        setNext(prev, follow)
+        len -= 1
+      }
+      prev = cur.asInstanceOf[Predecessor[A]]
+      cur = follow
+    }
+    this
+  }
+
+  def patchInPlace(from: Int, patch: collection.Seq[A], replaced: Int): this.type = {
+    ensureUnaliased()
+    val p = locate(from)
+    removeAfter(p, replaced `min` length - from)
```

Comment:
Precedence is not super-obvious here, especially with point-free backquoted `min`.